src/validation/validator.rs

```diff
@@ -0,0 +1,901 @@
+                                    uri,
+                                    format!("Cannot parse child certificate: {e}"),
+                                ));
+                            }
+                        };
+                    }
+                    "asa" => match Aspa::decode(object.bytes().clone(), strict) {
+                        Ok(aspa) => match aspa.process(resource_cert, true, |ee| {
+                            if crl.contains(ee.serial_number()) {
+                                Err(VerificationError::new(format!("ROA at {uri} was revoked"))
+                                    .into())
+                            } else {
+                                Ok(())
+                            }
+                        }) {
+                            Ok((_ee, aspa)) => {
+                                ca_cert.add_aspa(ValidatedAspa::make(uri, aspa));
+                            }
+                            Err(e) => {
+                                ca_cert.add_issue(ValidationIssue::with_uri_and_msg(
+                                    uri,
+                                    format!("Invalid ASPA: {e}"),
+                                ));
+                            }
+                        },
+                        Err(e) => {
+                            ca_cert.add_issue(ValidationIssue::with_uri_and_msg(
+                                uri,
+                                format!("Cannot parse aspa object: {e}"),
+                            ));
+                        }
+                    },
+                    "gbr" => ca_cert.add_ignored(IgnoredObjectInfo::new(
+                        uri,
+                        "Ghostbuster records are not yet validated.",
+                    )),
+                    _ => {
+                        let msg = format!("unsupported object type: {}", uri.extension());
+                        ca_cert.add_issue(ValidationIssue::with_uri_and_msg(uri, msg));
+                    }
+                }
+
+                validated += 1;
+            }
+        }
+
+        // Add the results for this CA certificate validation.
+        report.add_cert(ca_cert);
+
+        // recurse child certificates
+        for (cert, info) in child_certificates {
+            report.add_other(self.validate_tree_at(&cert, &info, offline, when));
+        }
+
+        report
+    }
+}
+
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub struct VisitedRepositories {
+    fetchers: HashMap<uri::Https, Arc<Fetcher>>,
+    data: HashMap<UriString, RepositoryData>,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct UriString(String);
+
+impl UriString {
+    pub fn as_str(&self) -> &str {
+        self.as_ref()
+    }
+
+    pub fn extension(&self) -> &str {
+        match self.as_str().rsplit_once('.') {
+            None => "",
+            Some((_left, extension)) => extension,
```

Comment:
Does the RFC mandate that the extension is lowercase or is that ensured elsewhere in krill-sync? Otherwise I could probably evade the check by creating something.ROA rather than something.roa seeing that further on you match on the lowercase version

Reply:
The RFC mandates lower case. If it is an unknown extension, validation for this object fails.